presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayAllocator.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.block;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import static java.lang.String.format;
+
+@NotThreadSafe
+public class ArrayAllocator
+{
+    private static final int DEFAULT_MAX_OUTSTANDING = 100;
+
+    private final Deque<byte[]> byteArrays = new ArrayDeque<>();
+    private final Deque<short[]> shortArrays = new ArrayDeque<>();
+    private final Deque<int[]> intArrays = new ArrayDeque<>();
+    private final Deque<long[]> longArrays = new ArrayDeque<>();
+    private final Deque<boolean[]> booleanArrays = new ArrayDeque<>();
+
+    private int outstandingArrays;
+    private final int maxOutstandingArrays;
+
+    public ArrayAllocator()
+    {
+        this(DEFAULT_MAX_OUTSTANDING);
+    }
+
+    public ArrayAllocator(int maxOutstandingArrays)
+    {
+        this.maxOutstandingArrays = maxOutstandingArrays;
+    }
+
+    public byte[] getByteArray(int positionCount)
+    {
+        checkState(outstandingArrays++ < maxOutstandingArrays, "Requested too many arrays: %s", outstandingArrays);
+        while (!byteArrays.isEmpty() && byteArrays.peek().length < positionCount) {
```

Comment:
Do we really want to discard all arrays that are smaller than the required size? There could be smaller requests coming in later. This policy makes the array sizes ever increasing but impossible to adjust down. I feel the policy in aria-research is more generic, what do you think?

Reply:
Just to be clear, the policy before is here:

https://github.com/prestodb/presto/blob/3c55db26282a7c83e5fe664940ee83d3e7baf2cc/presto-spi/src/main/java/com/facebook/presto/spi/block/IntArrayAllocator.java

Annotating it, it does the following:

1) If the list is empty, return a new array (this is the same)
2) Remove the last item from the list
3) If the size of this array is >= the desired size, then return it
4) Else, return a new array of the desired size

So it doesn't seem to really help the case you're describing, since we'll only select the smaller array by pure chance.  Did I understand correctly?